R/derive_param_extreme_event.R

```diff
@@ -0,0 +1,370 @@
+#'   the new observations.
+#'   1. The new observations are added to input dataset.
+#'
+#' @author Stefan Bundfuss Sophie Shapcott
+#'
+#' @return The input dataset with a new parameter indicating if and when an
+#'   event occurred
+#'
+#' @family der_prm_bds_findings
+#' @keywords der_prm_bds_findings
+#'
+#' @export
+#'
+#' @examples
+#' library(tibble)
+#' library(dplyr)
+#' library(lubridate)
+#'
+#' # Derive a new parameter for the first disease progression (PD)
+#' adsl <- tribble(
+#'   ~USUBJID, ~DTHDT,
+#'   "1",      ymd("2022-05-13"),
+#'   "2",      ymd(""),
+#'   "3",      ymd("")
+#' ) %>%
+#'   mutate(STUDYID = "XX1234")
+#'
+#' adrs <- tribble(
+#'   ~USUBJID, ~ADTC,        ~AVALC,
+#'   "1",      "2020-01-02", "PR",
+#'   "1",      "2020-02-01", "CR",
+#'   "1",      "2020-03-01", "CR",
+#'   "1",      "2020-04-01", "SD",
+#'   "2",      "2021-06-15", "SD",
+#'   "2",      "2021-07-16", "PD",
+#'   "2",      "2021-09-14", "PD"
+#' ) %>%
+#'   mutate(
+#'     STUDYID = "XX1234",
+#'     ADT = ymd(ADTC),
+#'     PARAMCD = "OVR",
+#'     PARAM = "Overall Response",
+#'     ANL01FL = "Y"
+#'   ) %>%
+#'   select(-ADTC)
+#'
+#' derive_param_extreme_event(
+#'   adrs,
+#'   dataset_adsl = adsl,
+#'   dataset_source = adrs,
+#'   filter_source = PARAMCD == "OVR" & AVALC == "PD",
+#'   order = vars(ADT),
+#'   new_var = AVALC,
+#'   true_value = "Y",
+#'   false_value = "N",
+#'   mode = "first",
+#'   set_values_to = vars(
+#'     PARAMCD = "PD",
+#'     PARAM = "Disease Progression",
+#'     ANL01FL = "Y",
+#'     ADT = ADT
+#'   )
+#' )
+#'
+#' # derive parameter indicating death
+#' derive_param_extreme_event(
+#'   dataset_adsl = adsl,
+#'   dataset_source = adsl,
+#'   filter_source = !is.na(DTHDT),
+#'   new_var = AVALC,
+#'   true_value = "Y",
+#'   false_value = "N",
+#'   mode = "first",
+#'   set_values_to = vars(
+#'     PARAMCD = "DEATH",
+#'     PARAM = "Death",
+#'     ANL01FL = "Y",
+#'     ADT = DTHDT
+#'   )
+#' )
+derive_param_extreme_event <- function(dataset = NULL,
+                                       dataset_adsl,
+                                       dataset_source,
+                                       filter_source,
+                                       order = NULL,
+                                       new_var = AVALC,
+                                       true_value = "Y",
+                                       false_value = "N",
+                                       mode = "first",
+                                       subject_keys = vars(STUDYID, USUBJID),
+                                       set_values_to,
+                                       check_type = "warning") {
+  # Check input parameters
+  filter_source <- assert_filter_cond(enquo(filter_source))
+  assert_vars(subject_keys)
+  assert_vars(order, optional = TRUE)
+  if (!is.null(order)) {
+    assert_data_frame(dataset_source, required_vars = vars(!!!subject_keys, !!!order))
```

Comment:
Should be `vars(!!!subject_keys, !!!extract_vars(order))` as we could have `order = vars(desc(AVISITN))`.